doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2308 @@
+It is implemented as an enum:
+```rust
+enum Option<T> {
+  Some(T),
+  None,
+}
+```
+
+The `<T>` is similar to the lifetime syntax we saw before; it means that `Option<T>` is a _generic type_; we'll dig into those soon.
+
+If we have a value of type `Option<T>` (or, any other enum, really), we can write code conditioned on the value's discriminant using _pattern matching_, which is accessed through the `match` expression:
+```rust
+let x: Option<u32> = ...;
+let y = match x {
+  Some(val) => val,  // If `x` is a `Some`, bind the value inside to `val`.
+  None => 42,  // If `x` is a `None`, do this instead.
+};
+```
+
+The key thing pattern matching gives us is the ability to inspect the union within an enum safely: the tag check is enforced by the compiler.
+
+Patterns are like expressions, forming a mini-language.
+If expressions build up a value by combining existing values, patterns do the opposite: they build up values by deconstructing values.
+In particular, a pattern, applied to an expression, performs the following operations:
+*   Checks that the expression's value actually matches that pattern.
+    (Note that type-checking doesn't go into this; patterns can't be used for conditioning on the type of an expression.)
+*   Optionally binds that expression's value to a name.
+*   Optionally recurs into subpatterns.
+
+Here's a few examples of patterns.
+Keep in mind the matching, binding, and recurrence properties of each.
+In general, patterns look like the value of the expression they match.
+
+*   `_`, an underscore[^85] pattern.
+    The match always succeeds, but it throws away the matched value.
+    This is the `_` in the equivalent of a `default:` case.
+*   `foo`, an identifier pattern.
+    This pattern is exactly like `_`, but it binds the matched value to its name.
+    This is the `val` in the `Some(val)` above.
+    This can also be used by itself as a default case that wants to do something with the matched-on value.
+    The binding can be made mutable by writing `Some(mut val)` instead.
+*   Any numeric literal, for a literal pattern.
+    This match compares the matched value against the literal value, and doesn't match anything.
+    These can also be inclusive ranges: `5..=16`[^86].
+*   `(pat1, pat2, /* etc */)`, a tuple pattern.
+    This match operates on tuple types, and always succeeds: it extracts the individual elements of a tuple, and applies them to the pattern's subpatterns.
+    In particular, the `()` pattern matches the unit value `()`.
+```rust
+let x: (u32, u32) = ...;
+match x {
+  (5, u) => { ... },  // Check that first element is five, 
+                      // bind the second element to `u`.
+  (u, _) => { ... },  // Bind the first element to `u`,
+                      // discard the second element.
+}
+
+let y: (u32, (u32, u32)) = ...;
+match y {
+  // All patterns can nest arbitrarily, like expressions.
+  (42, (u, _)) =>  { ... },
```

Comment:
I'd suggest not using {`...`} to denote code, because it's slightly confusing with `..` in tuple pattern matching. Also, I'd suggest putting `..` up above -- having it as a comment in the code means there are two places you lay out the possibilities. I missed it the first time.

Reply:
How would you feel about `s#...#/* ... */#g`?

Reply:
I think that would be clearer.

Reply:
Done.